File: Backend/app/schemas/verification.py
from pydantic import BaseModel, EmailStr, Field, field_validator

from app.schemas.token import Token


class VerifyEmailConfirm(BaseModel):
    otp: str = Field(min_length=6, max_length=6, pattern=r"^\d{6}$")


class ForgotPasswordRequest(BaseModel):
    email: EmailStr


class PasswordChangeBase(BaseModel):
    new_password: str = Field(min_length=8)

    @field_validator("new_password")
    @classmethod
    def validate_password_strength(cls, v: str) -> str:
        """Enforce password strength: min 8 chars, 1 upper, 1 lower, 1 digit, 1 special"""
        if len(v) < 8:
            raise ValueError("Password must be at least 8 characters")
        if not any(c.isupper() for c in v):
            raise ValueError("Password must contain at least one uppercase letter")
        if not any(c.islower() for c in v):
            raise ValueError("Password must contain at least one lowercase letter")
        if not any(c.isdigit() for c in v):
            raise ValueError("Password must contain at least one digit")
        if not any(not c.isalnum() for c in v):
            raise ValueError("Password must contain at least one special character")
        return v


class ResetPasswordConfirm(PasswordChangeBase):
    email: EmailStr
    otp: str = Field(min_length=6, max_length=6, pattern=r"^\d{6}$")


class ChangePasswordRequest(PasswordChangeBase):
    current_password: str


class GenericResponse(BaseModel):
    message: str


class VerifyEmailResponse(BaseModel):
    message: str
    tokens: "Token"

    class Config:
        from_attributes = True

File: movie_tracker_app/lib/models/auth/verification.dart
import 'token.dart';

class PasswordChangeBase {
  final String newPassword;
  final String confirmPassword;

  const PasswordChangeBase({
    required this.newPassword,
    required this.confirmPassword,
  });

  factory PasswordChangeBase.fromJson(Map<String, dynamic> json) {
    return PasswordChangeBase(
      newPassword: json['new_password'],
      confirmPassword: json['confirm_password'],
    );
  }

  Map<String, dynamic> toJson() => {
    'new_password': newPassword,
    'confirm_password': confirmPassword,
  };
}

class VerifyEmailConfirm {
  final String otp;

  const VerifyEmailConfirm({required this.otp});

  factory VerifyEmailConfirm.fromJson(Map<String, dynamic> json) {
    return VerifyEmailConfirm(otp: json['otp']);
  }

  Map<String, dynamic> toJson() => {'otp': otp};
}

class ForgotPasswordRequest {
  final String email;

  const ForgotPasswordRequest({required this.email});

  factory ForgotPasswordRequest.fromJson(Map<String, dynamic> json) {
    return ForgotPasswordRequest(email: json['email']);
  }

  Map<String, dynamic> toJson() => {'email': email};
}

class ResetPasswordConfirm extends PasswordChangeBase {
  final String email;
  final String otp;

  const ResetPasswordConfirm({
    required this.email,
    required this.otp,
    required super.newPassword,
    required super.confirmPassword,
  });

  factory ResetPasswordConfirm.fromJson(Map<String, dynamic> json) {
    return ResetPasswordConfirm(
      email: json['email'],
      otp: json['otp'],
      newPassword: json['new_password'],
      confirmPassword: json['confirm_password'],
    );
  }

  @override
  Map<String, dynamic> toJson() => {
    ...super.toJson(),
    'email': email,
    'otp': otp,
  };
}

class ChangePasswordRequest extends PasswordChangeBase {
  final String currentPassword;

  const ChangePasswordRequest({
    required this.currentPassword,
    required super.newPassword,
    required super.confirmPassword,
  });

  factory ChangePasswordRequest.fromJson(Map<String, dynamic> json) {
    return ChangePasswordRequest(
      currentPassword: json['current_password'],
      newPassword: json['new_password'],
      confirmPassword: json['confirm_password'],
    );
  }

  @override
  Map<String, dynamic> toJson() => {
    ...super.toJson(),
    'current_password': currentPassword,
  };
}

class GenericResponse {
  final String message;

  const GenericResponse({required this.message});

  factory GenericResponse.fromJson(Map<String, dynamic> json) {
    return GenericResponse(message: json['message']);
  }

  Map<String, dynamic> toJson() => {'message': message};
}

class VerifyEmailResponse {
  final String message;
  final Token tokens;

  const VerifyEmailResponse({required this.message, required this.tokens});

  factory VerifyEmailResponse.fromJson(Map<String, dynamic> json) {
    return VerifyEmailResponse(
      message: json['message'],
      tokens: Token.fromJson(json['tokens']),
    );
  }

  Map<String, dynamic> toJson() => {
    'message': message,
    'tokens': tokens.toJson(),
  };
}